README.md
# indexed_gzip
 

[![PyPi version](https://img.shields.io/pypi/v/indexed_gzip.svg)](https://pypi.python.org/pypi/indexed_gzip/) [![Anaconda version](https://anaconda.org/conda-forge/indexed_gzip/badges/version.svg)](https://anaconda.org/conda-forge/indexed_gzip/)![Test status](https://github.com/pauldmccarthy/indexed_gzip/actions/workflows/master.yaml/badge.svg)


 *Fast random access of gzip files in Python*


 * [Overview](#overview)
 * [Installation](#installation)
 * [Usage](#usage)
 * [Using with `nibabel`](#using-with-nibabel)
 * [Index import/export](#index-import-export)
 * [Write support](#write-support)
 * [Performance](#performance)
 * [Acknowledgements](#acknowledgements)
 * [License](#license)


## Overview


The `indexed_gzip` project is a Python extension which aims to provide a
drop-in replacement for the built-in Python `gzip.GzipFile` class, the
`IndexedGzipFile`.


`indexed_gzip` was written to allow fast random access of compressed
[NIFTI](http://nifti.nimh.nih.gov/) image files (for which GZIP is the
de-facto compression standard), but will work with any GZIP file.
`indexed_gzip` is easy to use with `nibabel` (http://nipy.org/nibabel/).


The standard `gzip.GzipFile` class exposes a random access-like interface (via
its `seek` and `read` methods), but every time you seek to a new point in the
uncompressed data stream, the `GzipFile` instance has to start decompressing
from the beginning of the file, until it reaches the requested location.


An `IndexedGzipFile` instance gets around this performance limitation by
building an index, which contains *seek points*, mappings between
corresponding locations in the compressed and uncompressed data streams. Each
seek point is accompanied by a chunk (32KB) of uncompressed data which is used
to initialise the decompression algorithm, allowing us to start reading from
any seek point. If the index is built with a seek point spacing of 1MB, we
only have to decompress (on average) 512KB of data to read from any location
in the file.


## Intended use


You may find `indexed_gzip` useful if you need to read from large GZIP files.
A major advantage of `indexed_gzip` is that it will work with any GZIP file.
However, if you have control over the creation of your GZIP files, you may
wish to consider some alternatives:

 * [`mgzip`](https://github.com/vinlyx/mgzip/) provides an accelerated
   GZIP compression and decompression library.
 * Compression formats other than GZIP, such as `bzip2` and `xz`, have better
   support for random access.


## Installation


`indexed_gzip` is available on [PyPi](https://pypi.python.org/pypi) - to
install, simply type:
```sh
pip install indexed_gzip
```


You can also install `indexed_gzip` from conda-forge:

```sh
conda install -c conda-forge indexed_gzip
```

To compile `indexed_gzip`, make sure you have [cython](http://cython.org/)
installed (and `numpy` if you want to compile the tests), and then run:
```sh
python setup.py develop
```


To run the tests, type the following; you will need `numpy`, `nibabel`,
`pytest`, `pytest-cov`, and `coverage` installed:

```sh
python -m indexed_gzip.tests
```

## Usage


You can use the `indexed_gzip` module directly:


```python
import indexed_gzip as igzip

# You can create an IndexedGzipFile instance
# by specifying a file name, or an open file
# handle. For the latter use, the file handle
# must be opened in read-only binary mode.
# Write support is currently non-existent.
myfile = igzip.IndexedGzipFile('big_file.gz')

some_offset_into_uncompressed_data = 234195

# The index will be automatically
# built on-demand when seeking.
myfile.seek(some_offset_into_uncompressed_data)
data = myfile.read(1048576)
```


## Using with in-memory data

You can use `indexed_gzip` with any Python file-like object. For example:

```python

import io
import indexed_gzip as igzip

# Load some gzip data from somewhere
with open('my_file.gz') as f:
    data = f.read()

# Create an IndexedGzipFile based on the
# in-memory data buffer
gzf = igzip.IndexedGzipFile(fileobj=io.BytesIO(data))
uncompressed = gzf.read(1048576)
```


## Using with `nibabel`


You can use `indexed_gzip` with `nibabel`. `nibabel` >= 2.3.0 will
automatically use `indexed_gzip` if it is present:


```python
import nibabel as nib

image = nib.load('big_image.nii.gz')
```


If you are using `nibabel` 2.2.x, you need to explicitly set the
`keep_file_open` flag:


```python
import nibabel as nib

image = nib.load('big_image.nii.gz', keep_file_open='auto')
```


To use `indexed_gzip` with `nibabel` 2.1.0 or older, you need to do a little
more work:


```python
import nibabel      as nib
import indexed_gzip as igzip

# Here we are using 4MB spacing between
# seek points, and using a larger read
# buffer (than the default size of 16KB).
fobj = igzip.IndexedGzipFile(
    filename='big_image.nii.gz',
    spacing=4194304,
    readbuf_size=131072)

# Create a nibabel image using
# the existing file handle.
fmap = nib.Nifti1Image.make_file_map()
fmap['image'].fileobj = fobj
image = nib.Nifti1Image.from_file_map(fmap)

# Use the image ArrayProxy to access the
# data - the index will automatically be
# built as data is accessed.
vol3 = image.dataobj[:, :, :, 3]
```


## Index import/export


If you have a large file, you may wish to pre-generate the index once, and
save it out to an index file:


```python
import indexed_gzip as igzip

# Load the file, pre-generate the
# index, and save it out to disk.
fobj = igzip.IndexedGzipFile('big_file.gz')
fobj.build_full_index()
fobj.export_index('big_file.gzidx')
```


The next time you open the same file, you can load in the index:


```python
import indexed_gip as igzip
fobj = igzip.IndexedGzipFile('big_file.gz', index_file='big_file.gzidx')
```


## Write support


`indexed_gzip` does not currently have any support for writing. Currently if
you wish to write to a file, you will need to save the file by alternate means
(e.g.  via `gzip` or `nibabel`), and then re-create a new `IndexedGzipFile`
instance.  For example:


```python

import nibabel as nib

# Load the entire image into memory
image = nib.load('big_image.nii.gz')
data = image.get_data()

# Make changes to the data
data[:, :, :, 5] *= 100

# Save the image using nibabel
nib.save(data, 'big_image.nii.gz')

# Re-load the image
image = nib.load('big_image.nii.gz')
```


## Performance


A small [test script](indexed_gzip/tests/benchmark.py) is included with
`indexed_gzip`; this script compares the performance of the `IndexedGzipFile`
class with the `gzip.GzipFile` class. This script does the following:

  1. Generates a test file.

  2. Generates a specified number of seek locations, uniformly spaced
     throughout the test file.

  3. Randomly shuffles these locations

  4. Seeks to each location, and reads a chunk of data from the file.


This plot shows the results of this test for a few compresed files of varying
sizes, with 500 seeks:


![Indexed gzip performance](./performance.png)


## Acknowledgements


The `indexed_gzip` project is based upon the `zran.c` example (written by Mark
Alder) which ships with the [zlib](http://www.zlib.net/) source code.


`indexed_gzip` was originally inspired by Zalan Rajna's (@zrajna)
[zindex](https://github.com/zrajna/zindex) project:

    Z. Rajna, A. Keskinarkaus, V. Kiviniemi and T. Seppanen
    "Speeding up the file access of large compressed NIfTI neuroimaging data"
    Engineering in Medicine and Biology Society (EMBC), 2015 37th Annual
    International Conference of the IEEE, Milan, 2015, pp. 654-657.

    https://sourceforge.net/projects/libznzwithzindex/


Initial work on `indexed_gzip` took place at
[Brainhack](http://www.brainhack.org/) Paris, at the Institut Pasteur,
24th-26th February 2016, with the support of the
[FMRIB Centre](https://www.ndcn.ox.ac.uk/divisions/fmrib/), at the
University of Oxford, UK.

Many thanks to the following contributors (listed chronologically):

 - Zalan Rajna (@zrajna): Bug fixes (#2)
 - Martin Craig (@mcraig-ibme): Porting `indexed_gzip` to Windows (#3)
 - Chris Markiewicz (@effigies): Option to drop file handles (#6)
 - Omer Ozarslan (@ozars): Index import/export (#8)
 - @DarioDaF: Windows overflow bug (#30)
 - Sławomir Zborowski (@szborows): `seek_points` method (#35), README fixes
   (#34)
 - Ashwin Ramaswami (@epicfaace): Support for in-memory file objects (#55),
   bug fixes (#63, #64, #65).
 - Michał Górny (@mgorny): Remove hard dependency on `nibabel` from test suite
   (#78).
 - Alexander Gorban (@alexgorban) Fix memory leak (#82, #83).
 - Maximilian Knespel (@mxmlnkn) Change default read buffer size to
   improve performance (#90).


## License


`indexed_gzip` inherits the [zlib](http://www.zlib.net) license, available for
perusal in the [LICENSE](LICENSE) file.


CHANGELOG.md
# `indexed_gzip` changelog


## 1.8.0 (Under development)


* Changes to allow linking against
  [zlib-ng](https://github.com/zlib-ng/zlib-ng/) (#107).
* Simplified C interface `cindexed_gzip`, and machinery for compiling as a
  static or shared C library (#107).


## 1.7.1 (March 31st 2023)


* Small change to the `IndexedGzipFile` class so that it accepts file-likes
  which do not implement `fileno()` (#118).


## 1.7.0 (September 12th 2022)


* Changes to allow an index to be built from file-likes which don't support
  `seek()` or `tell()` operations (#103, #105).


## 1.6.13 (April 14th 2022)


* Changed the default read buffer size used by the `IndexedGzipFile` class
  (#90).
* Update to wheel building procedure (#92,#93,#95,#96,#97,#98,#99).
* Aded `pyproject.toml` to declare build-time requirements (#94).


## 1.6.4 (October 18th 2021)


* Fixed a bug related to buffering input data, which was causing a spurious
  `CrcError` (#80, #87).



## 1.6.3 (September 14th 2021)


* Relaxed `mode` check when creating an `IndexedGzipFile` from an open file
  handle - `fileobj.mode` may now be `'r'` or `'rb'` (or may not exist at all)
  (#85, #86).


## 1.6.2 (September 2nd 2021)


* Fixed a memory leak when initialising decompression / inflation (#82, #83).
* Added file name to exception messages when possible, to assist in diagnosing
  errors (#84).


## 1.6.1 (May 25th 2021)


* Tests requiring `nibabel` are now skipped, rather than causing failure (#78).


## 1.6.0 (May 23rd 2021)


* Python 2.7 wheels for Windows are no longer being built (#71, #73).
* A backwards-compatible change to the index file format, to accommodate seek
  points at stream boundaries. Index files created with older versions of
  `indexed_gzip` can still be loaded, but index files created with
  `indexed_gzip` 1.6.0 cannot be loaded by older versions of `indexed_gzip`
  (#75).
* CRC and size validation of uncompressed data is now performed by default,
  on the first pass through a GZIP file. This can be disabled by setting the
  new `skip_crc_check` argument to `False` when creating an
  `IndexedGzipFile`. Validation is not performed when an existing index is
  imported from file (#72).
* Null padding bytes at the end of a GZIP file, or in between GZIP streams,
  are now skipped over (#69, #70, #72).
* Seek points are now created at the beginning of every GZIP stream, in files
  containing concatenated streams (#72).


## 1.5.3 (March 23rd 2021)


* Restored wheel building for Python 2.7 on 32 bit Windows (#64, #66).
* Now building wheels for `aarch64` on Linux, and `amd64`/`universal2` on
  macOS (#66).
* Fixed some un-initialised pointers (#65).
* Fixed a bug in the use of `PyErr_Occurred` (#63).


## 1.5.2 (March 19th 2021)


* Not providing binary wheels for Python 2.7 on 32 bit Windows (#61).


## 1.5.1 (March 19th 2021)


* Minor adjustments to some unit test (#60).


## 1.5.0 (March 19th 2021)


* Added support for in-memory file-like objects (#55).
* Fixed a bug whereby a segmentation fault could occur if an `IndexedGzipFile`
  was created with a path to a non-existent file (#56).


## 1.4.0 (January 2nd 2021)


* Fixed a bug in the pickling/copying logic in the `IndexedGzipFile` class
  (#50, #51)
* New `indexed_gzip.open` function, which just creates and returns an
  `IndexedGzipFile`.
* When creating an `IndexedGzipFile`, the first argument (`filename`) may
  be either a file name, or an open file handle (#49, #53).
* Migrated CI testing and building to Github Actions (#52).
* Binary wheels for Windows now have ZLIB statically compiled in as part
  of the wheel, so ZLIB no longer needs to be installed (#43, #52).


## 1.3.3 (November 14th 2020)


* Adjusted the `ZranError` exception type to sub-class from `IOError`, to
  ease support for `nibabel`.
* Fixed a bug related to concatenated GZIP files which would occur when
  the read buffer ran out of space at the point where a stream boundary
  occurred.


## 1.3.2 (June 30th 2020)


* Adjusted the `NoHandleError` and `NotCoveredError` types to sub-class from
  `ValueError`, to preserve backwards compatibility with older versions of
  `nibabel`.


## 1.3.1 (June 25th 2020)


* Reverted the error type raised by the `IndexedGzipFile.seek` to `ValueError`,
  as `nibabel` assumes that the `seek` method of file objects raise a
  `ValueError` if `SEEK_END` is not supported.


## 1.3.0 (June 24th 2020)


* The `IndexedGzipFile.seek` method now accepts seeking from the end of
  uncompressed stream via `SEEK_END`, as long as the index has been built (#37).


## 1.2.0 (May 12th 2020)


* New `IndexedGzipFile.seek_points` method, which returns the compressed and
  uncompressed seek point locations.


## 1.1.0 (April 20th 2020)


* `IndexedGzipFile` objects are now picklable, as long as they are created with
  the default setting of `drop_handles=True` and with a `filename` and not an
  open `fileobj` (#28, #31).
* Changed the return type of `zran_tell` from `long` to `uint64_t`, because the
  former is not guaranteed to be 64 bit (#29, #30).
* Changed the `zran_index_t.compressed_size` and `uncompressed_size` fields from
  `size_t` to `uint64_t` because the former is not guaranteed to be 64 bit.


## 1.0.0 (February 21st 2020)


* Removed the deprecated `fid` argument to the `IndexedGzipFile` constructor.
* Removed the `SafeIndexedGzipFile`.


## 0.8.10 (May 15th 2019)


* Fix to package metadata.


## 0.8.9 (May 14th 2019)


* The `IndexedGzipFile.import_index` method and `zran_import_index` function
  can handle index files which do not contain any index points (#18).


## 0.8.8 (November 22nd 2018)

* Fixed bug affecting files which were an exact multiple of the read buffer
  size (#15).


## 0.8.7 (August 3rd 2018)

* Internal changes to how file handles are managed, to improve Windows
  compatibility.
* The `_IndexedGzipFile.read` method now acquires a single file handle, rather
  than opening/closing multiple handles across calls to `zran_read`.



cindexed_gzip/zran.h
#ifndef __ZRAN_H__
#define __ZRAN_H__


#ifdef __cplusplus
extern "C" {
#endif


/*
 * The zran module is an adaptation of the zran example, written by Mark
 * Alder, which ships with the zlib source code. It allows the creation
 * of an index into a compressed file, which is used to improve the speed
 * of random seek/read access to the uncompressed data.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#ifdef ZRAN_SUPPORT_PYTHON
#define PY_SSIZE_T_CLEAN
#include <Python.h>
#else
#define PyObject void
#endif

struct _zran_index;
struct _zran_point;


typedef struct _zran_index zran_index_t;
typedef struct _zran_point zran_point_t;

/*
 * These values may be passed in as flags to the zran_init function.
 * They are specified as bit-masks, rather than bit locations.
 */
enum {
  ZRAN_AUTO_BUILD     = 1,
  ZRAN_SKIP_CRC_CHECK = 2,
};


/*
 * Struct representing the index. None of the fields in this struct
 * should ever need to be accessed or modified directly.
 */
struct _zran_index {

    /*
     * Handle to the compressed file.
     */
    FILE *fd;

    /*
     * Handle to the compressed file object.
     */
    PyObject *f;

    /*
     * Size of the compressed file. This
     * is calculated in zran_init.
     */
    uint64_t compressed_size;

    /*
     * Size of the uncompressed data. This is
     * only updated when it becomes known.
     */
    uint64_t uncompressed_size;

    /*
     * Spacing size in bytes, relative to the
     * uncompressed data stream, between adjacent
     * index points.
     */
    uint32_t spacing;

    /*
     * Number of bytes of uncompressed data to store
     * for each index point. This must be a minimum
     * of 32768 bytes.
     */
    uint32_t window_size;

    /*
     * Base2 logarithm of the window size - it
     * is needed to initialise zlib inflation.
     */
    uint32_t log_window_size;

    /*
     * Size, in bytes, of buffer used to store
     * compressed data read from disk.
     */
    uint32_t readbuf_size;

    /*
     * Number of index points that have been created.
     */
    uint32_t npoints;

    /*
     * Number of index points that can be stored -
     * i.e. the amount allocated to the "list" field.
     */
    uint32_t size;

    /*
     * List of index points.
     */
    zran_point_t *list;

    /*
     * Most recently requested seek/read
     * location into the uncompressed data
     * stream - this is used to keep track
     * of where the calling code thinks it
     * is in the (uncompressed) file.
     */
    uint64_t uncmp_seek_offset;

    /*
     * Flags passed to zran_init
     */
    uint16_t flags;

    /*
     * All of the fields after this point are used
     * by the internal _zran_inflate function.
     */

    /*
     * Reference to a file input
     * buffer of size readbuf_size.
     */
    uint8_t *readbuf;

    /*
     * An offset into readbuf.
     */
    uint32_t readbuf_offset;

    /*
     * The current end of the readbuf contents.
     */
    uint32_t readbuf_end;

    /*
     * Current offsets into the uncompressed and
     * compressed data streams.
     */
    uint64_t inflate_cmp_offset;
    uint64_t inflate_uncmp_offset;

    /*
     * Uncompressed offset at the point that the
     * last GZIP stream ended. This is updated as
     * more data is read and uncompressed, and
     * used to determine whether the CRC/size
     * check for the current stream has already
     * been performed.
     */
    uint64_t last_stream_ended;

    /*
     * CRC-32 checksum and size (number of
     * bytes, modulo 2^32) of the uncompressed
     * data in the current gzip stream, not
     * used if ZRAN_SKIP_CRC_CHECK is active.
     * The CRC and size are incrementally
     * calculated as data is read in. When the
     * end of a gzip stream is reached, the
     * calculated CRC and size are compared
     * against the CRC and size in the gzip
     * footer, and an error is returned if
     * they don't match.
     */
    uint32_t stream_crc32;
    uint32_t stream_size;
    uint8_t  validating;
};


/*
 * Struct representing a single seek point in the index.
 */
struct _zran_point {


    /*
     * Location of this point in the compressed data
     * stream. This is the location of the first full
     * byte of compressed data - if  the compressed
     * and uncompressed locations are not byte-aligned,
     * the bits field below specifies the bit offset.
     */
    uint64_t  cmp_offset;

    /*
     * Corresponding location of this point
     * in the uncompressed data stream.
     */
    uint64_t  uncmp_offset;

    /*
     * If this point is not byte-aligned, this specifies
     * the number of bits, in the compressed stream,
     * back from cmp_offset, that the uncompressed data
     * starts.
     */
    uint8_t   bits;

    /*
     * Chunk of uncompressed data preceding this point.
     * This is required to initialise decompression from
     * this point onward.
     */
    uint8_t  *data;
};


/*
 * Initialise a zran_index_t struct for use with the given file.
 *
 * Passing in 0 for the spacing, window_size and readbuf_size arguments
 * will result in the following values being used:
 *
 *    spacing:      1048576
 *    window_size:  32768
 *    readbuf_size: 16384
 *
 * The read buffer must be at least the maximum expected size of a GZIP
 * header. GZIP headers have a minimum size of 10 bytes, but there is no upper
 * bound on their size, so using a very small read buffer would be unwise.  In
 * the case of concatenated GZIP streams, the read buffer must be at least big
 * enough to accommodate a GZIP footer of one stream, padding bytes in between
 * two streams, and the GZIP header of the next stream.
 *
 * The flags argument is a bit mask used to control the following options:
 *
 *     ZRAN_AUTO_BUILD:     Build the index automatically on demand.
 *
 *     ZRAN_SKIP_CRC_CHECK: Do not perform a CRC32 and file size check
 *                          when the end of a GZIP stream is reached.
 *                          This flag is automatically set when an index
 *                          is imported from file using zran_import_index.
 */
int  zran_init(
  zran_index_t *index,        /* The index                                  */
  FILE         *fd,           /* Open handle to the compressed file         */
  PyObject     *f,            /* Open handle to the compressed file object  */
  uint32_t      spacing,      /* Distance in bytes between
                                 index seek points                          */
  uint32_t      window_size,  /* Number of uncompressed bytes
                                 to store with each point                   */
  uint32_t      readbuf_size, /* Number of bytes to read at a time          */
  uint16_t      flags         /* Flags controlling index behaviour          */
);


/*
 * Frees the memory use by the given index. The zran_index_t struct
 * itself is not freed.
 */
void zran_free(
  zran_index_t *index /* The index */
);

/*
 * Return codes for zran_build_index.
 */
enum {
    ZRAN_BUILD_INDEX_OK        =  0,
    ZRAN_BUILD_INDEX_FAIL      = -1,
    ZRAN_BUILD_INDEX_CRC_ERROR = -2,
};


/*
 * (Re-)Builds the index to cover the given range, which must be
 * specified relative to the compressed data stream. Pass in 0
 * for both offsets to re-build the full index.
 *
 * Returns ZRAN_BUILD_INDEX_OK on success, ZRAN_BUILD_INDEX_CRC_ERROR
 * if a CRC error is detected in a GZIP stream, or ZRAN_BUILD_INDEX_FAIL
 * if some other type of error occurs.
 */
int zran_build_index(
  zran_index_t *index, /* The index */
  uint64_t      from,  /* Build the index from this point */
  uint64_t      until  /* Build the index to this point   */
);


/* Return codes for zran_seek. */
enum {
    ZRAN_SEEK_CRC_ERROR       = -2,
    ZRAN_SEEK_FAIL            = -1,
    ZRAN_SEEK_OK              =  0,
    ZRAN_SEEK_NOT_COVERED     =  1,
    ZRAN_SEEK_EOF             =  2,
    ZRAN_SEEK_INDEX_NOT_BUILT =  3
};


/*
 * Seek to the specified offset in the uncompressed data stream.
 * If the index does not currently cover the offset, and it was
 * created with the ZRAN_AUTO_BUILD flag, the index is expanded
 * to cover the offset.
 *
 * Seeking from the end of the uncompressed stream (using SEEK_END)
 * is only possible if the index fully covers the file.
 *
 * Returns:
 *    - ZRAN_SEEK_OK for success.
 *
 *    - ZRAN_SEEK_INDEX_NOT_BUILT if SEEK_END is used, and the index
 *      does not fully cover the file.
 *
 *    - ZRAN_SEEK_NOT_COVERED to indicate that the index does not
 *      cover the requested offset (will never happen if
 *      ZRAN_AUTO_BUILD is active).
 *
 *    - ZRAN_SEEK_EOF to indicate that the requested offset
 *      is past the end of the uncompressed stream.
 *
 *    - ZRAN_SEEK_CRC_ERROR to indicate that the CRC or file size
 *      stored in the footer of a GZIP stream does not match the
 *      data.
 *
 *    - ZRAN_SEEK_FAIL to indicate failure of some sort.
 */
int zran_seek(
  zran_index_t  *index,   /* The index                       */
  int64_t        offset,  /* Uncompressed offset to seek to  */
  uint8_t        whence,  /* SEEK_SET, SEEK_CUR, or SEEK_END */
  zran_point_t **point    /* Optional place to store
                             corresponding zran_point_t      */
);

/*
 * Returns the current seek location in the uncompressed data stream
 * (just returns zran_index_t.uncmp_seek_offset).
 */
uint64_t zran_tell(
  zran_index_t *index /* The index */
);


/* Return codes for zran_read. */
enum {
    ZRAN_READ_NOT_COVERED = -1,
    ZRAN_READ_EOF         = -2,
    ZRAN_READ_FAIL        = -3,
    ZRAN_READ_CRC_ERROR   = -4
};

/*
 * Read len bytes from the current location in the uncompressed
 * data stream, storing them in buf. If the index was created with
 * the ZRAN_AUTO_BUILD flag, it is expanded as needed.
 *
 * Returns:
 *   - Number of bytes read for success, or one of the following codes,
 *     all of which are negative.
 *
 *   - ZRAN_READ_NOT_COVERED to indicate that the index does not
 *     cover the requested region (will never happen if
 *     ZRAN_AUTO_BUILD is active).
 *
 *   - ZRAN_READ_EOF to indicate that the read could not be completed
 *     because the current uncompressed seek point is at EOF.
 *
 *   - ZRAN_SEEK_CRC_ERROR to indicate that the CRC or file size
 *     stored in the footer of a GZIP stream does not match the
 *     data.
 *
 *   - ZRAN_READ_FAIL to indicate that the read failed for some reason.
 */
int64_t zran_read(
  zran_index_t  *index, /* The index                 */
  void          *buf,   /* Buffer to store len bytes */
  uint64_t       len    /* Number of bytes to read   */
);

/*
 * Identifier and version number for index files created by zran_export_index,
 * defined in zran.c.
 */
extern const char    ZRAN_INDEX_FILE_ID[];
extern const uint8_t ZRAN_INDEX_FILE_VERSION;

/* Return codes for zran_export_index. */
enum {
    ZRAN_EXPORT_OK          =  0,
    ZRAN_EXPORT_WRITE_ERROR = -1
};

/*
 * Export current index data to given file. This exported file later can be
 * used to rebuild index without needing to going through the file again.
 *
 * See zran_import_index for importing.
 *
 * A zran index file is a binary file which has the following header
 * structure. All fields are assumed to be stored with little-endian
 * ordering:
 *
 * | Offset | Length | Description                           |
 * | 0      | 5      | File header (ascii, GZIDX)            |
 * | 5      | 1      | Version (uint8, currently 1)          |
 * | 6      | 1      | Reserved (uint8, currently must be 0) |
 * | 7      | 8      | Compressed file size  (uint64)        |
 * | 15     | 8      | Uncompressed file size (uint64)       |
 * | 23     | 4      | Index point spacing (uint32)          |
 * | 27     | 4      | Index window size W (uint32)          |
 * | 31     | 4      | Number of index points (uint32)       |
 *
 * The header is followed by the offsets for each index point:
 *
 * | Offset | Length | Description                              |
 * | 0      | 8      | Compressed offset for point 0 (uint64)   |
 * | 8      | 8      | Uncompressed offset for point 0 (uint64) |
 * | 16     | 1      | Bit offset for point 0 (uint8)           |
 * | 17     | 1      | Data flag - 1 if point has window data,  |
 * |        |        | 0 otherwise (uint8, added in file format |
 * |        |        | version 1)                               |
 * | ...    | ...    | ...                                      |
 * | N*18   | 8      | Compressed offset for point N (uint64)   |
 * | ...    | ...    | ...                                      |
 *
 * Finally the window data for all index points that have data is
 * concatenated (W represents the index window size):
 *
 * | Offset | Length | Description                                 |
 * | 0      | W      | Window data for first index point with data |
 * | ...    | ...    | ...                                         |
 * | N*W    | W      | Window data for Nth index point with data   |
 *
 * Returns:
 *   - ZRAN_EXPORT_OK for success.
 *
 *   - ZRAN_EXPORT_WRITE_ERROR to indicate an error from writing to underlying
 *     file.
 */
int zran_export_index(
  zran_index_t  *index, /* The index                         */
  FILE          *fd,    /* Open handle to export file        */
  PyObject      *f      /* Open handle to export file object */
);


/* Return codes for zran_import_index. */
enum {
    ZRAN_IMPORT_OK                  =  0,
    ZRAN_IMPORT_FAIL                = -1,
    ZRAN_IMPORT_EOF                 = -2,
    ZRAN_IMPORT_READ_ERROR          = -3,
    ZRAN_IMPORT_INCONSISTENT        = -4,
    ZRAN_IMPORT_MEMORY_ERROR        = -5,
    ZRAN_IMPORT_UNKNOWN_FORMAT      = -6,
    ZRAN_IMPORT_UNSUPPORTED_VERSION = -7
};


/*
 * Import current index from the given file.  index must have been initialized
 * by zran_init function before calling this function, as it is not supported
 * importing into an uninitialised zran_index_t struct. Existing index will be
 * overwritten including spacing and window_size values, whereas values of
 * readbuf_size and flags will be kept.
 *
 * Updating an index file is not supported currently. To update an index file,
 * first import it, create new checkpoints, and then export it again.
 *
 * CRC validation of uncompressed data from an imported index is not currently
 * supported - this function will enable the ZRAN_SKIP_CRC_CHECK flag on the
 * given zran_index_t struct.
 *
 * See zran_export_index for exporting.
 *
 * Returns:
 *   - ZRAN_IMPORT_OK for success.
 *
 *   - ZRAN_IMPORT_FAIL general errors.
 *
 *   - ZRAN_IMPORT_EOF to indicate unexpected end-of-file.
 *
 *   - ZRAN_IMPORT_READ_ERROR to indicate error while reading file.
 *
 *   - ZRAN_IMPORT_OVERFLOW to indicate overflow while reading compressed and
 *     uncompressed size attributes. This shouldn't be a problem for x64
 *     processors.
 *
 *   - ZRAN_IMPORT_INCONSISTENT to indicate compressed size, or uncompressed
 *     size if known,   of the index file is inconsistent with the loaded
 *     compressed file.
 *
 *   - ZRAN_IMPORT_MEMORY_ERROR to indicate failure to allocate memory for new
 *     index. This typically result from out-of-memory.
 *
 *   - ZRAN_IMPORT_UNKNOWN_FORMAT to indicate given file is of unknown format.
 *
 *   - ZRAN_IMPORT_UNSUPPORTED_VERSION to indicate that the file has a version
 *     which is too new for this version of indexed_gzip to parse.
 */
int zran_import_index(
  zran_index_t  *index, /* The index                         */
  FILE          *fd,    /* Open handle to import file        */
  PyObject      *f      /* Open handle to import file object */
);

#ifdef __cplusplus
}
#endif

#endif /* __ZRAN_H__ */
